File: python/knapsack/knapsack.py
"""Brute force the knapsack problem."""
import functools


@functools.cache
def maximum(cap: int, items: tuple[tuple[int, int]]) -> int:
    """Solve for the maximum backpack value."""
    options = []
    for i, (weight, val) in enumerate(items):
        if weight > cap:
            continue
        n_items = items[:i] + items[i + 1:]
        options.append(val + maximum(cap - weight, n_items))
    return max(options) if options else 0


def maximum_value(maximum_weight: int, items: list[dict[str, int]]) -> int:
    """Return the maximum backpack value."""
    item_tuple = tuple((i["weight"], i["value"]) for i in items)
    return maximum(maximum_weight, item_tuple)

File: go/knapsack/knapsack.go
package knapsack

// Item represents one item which can be carried.
type Item struct {
	Weight, Value int
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Knapsack takes in a maximum carrying capacity and a collection of items
// and returns the maximum value that can be carried by the knapsack
// given that the knapsack can only carry a maximum weight given by maximumWeight
func Knapsack(maximumWeight int, items []Item) int {
	if len(items) == 0 {
		return 0
	}
	one := items[0]
	rest := items[1:]
	if one.Weight > maximumWeight {
		return Knapsack(maximumWeight, rest)
	}
	return max(
		one.Value+Knapsack(maximumWeight-one.Weight, rest),
		Knapsack(maximumWeight, rest),
	)
}
